Help page for repo templates (#2110)
education/classroom: docs/using-template-repos-for-assignments.md

Importing starter code for assignments is a very useful feature and with GitHub recently shipping the template repositories, we can leverage them in Classroom to make importing starter code much easier, and faster!

### How to use template repositories for an assignment?

> ![Searches for repository, then selects the option of Importing starter code with template repositories](https://user-images.githubusercontent.com/1490434/61005360-10631080-a336-11e9-843b-ddb3e4f011ca.gif)

Reviewer: I don't think we should use this gif. The random letter typing (rather than a valid repository) and quick movements might get confusing for teachers and first-time users.

When you pick your template repository as a starter code repository, you will now get two options:

  * Import starter code using a template repository.
  * Import starter code using source importer.

>   * After selecting either of the options, the following steps are the same as before. More information on this can be found on our [Videos page](/videos)

Reviewer: This shouldn't be in a bullet point. Looks a little weird since you're saying "two options" and then have three bullet points.
> the following steps are the same as before
This might not be the best choice of words since you're assuming the people reading this are existing users. New users will definitely be reading this too.

Author: Oops, i thought i fixed that. Good catch!

> Note: If the picked starter code repository is not a template repository we will try to make it into one for you. If that fails, you can do it manually from the settings page of the repository on GitHub. More details on this page: [https://help.github.com/en/articles/creating-a-template-repository](https://help.github.com/en/articles/creating-a-template-repository)

Reviewer: This isn't true _yet_ (#2080) and might take some time if we're going to add a background javascript check on the form. Probably best to leave this out for now (keep the instructions to make it a template though)

### What are the advantages of using template repositories?

* Importing starter code is faster, as it uses the newer and faster [template repositories API](https://developer.github.com/v3/repos/#create-repository-using-a-repository-template).

> * Template repositories can be really useful if only the lastest files in the starter code repository are important and you do not want the entire commit history.

Reviewer: This seems more like a disadvantage and might be best to leave it in the "what do you miss out" section

### What do you miss out on when you use template repositories?\*
As you don’t get the complete commit history, you miss out on the following features:

* All the original tags in the repositories.
* Since repositories created from templates are a clean slate, adding them as remotes on original repository and pushing updates to them is not possible.

> ### Recommended option for your assignment

Reviewer: I feel like this section is over-complicated and mentions a lot of specifics. Can I suggest something like this?
```
Template repositories are the best option in _most_ cases. You may not want to use it if:
- You plan on pushing updates to the starter code repository _after_ releasing the assignment
- You care about the full commit history of the starter code repository
```
```

* A stand-alone homework assignment with starter code, or with just a README, or assignment instructions is a perfect fit for using template repositories.
* If you’re planning to create an assignment for a group project or a long project with multiple phases, you can either use separate branches for each phase and use repo templates, or you can stick to the importer.
* If you plan on pushing changes to the student repos for any updates, using the existing importer might be a better option

> \* the template repos API is still under a preview and might change, but we will keep this page updated if any of those changes affect your experience on classroom.

Reviewer: ```suggestion
\* The template repositories API is still under a preview and might change.  We will keep this page updated if any of those changes affect your experience on classroom.
```

Importing starter code for assignments is a very useful feature and with GitHub recently shipping the template repositories, we can leverage them in Classroom to make importing starter code much easier and faster!

### How to use template repositories for an assignment?

> #### Before you begin

Reviewer: I feel like this section is kind-of a given (e.g. set up classroom) and unnecessary. @d12 what do you think?

> #### Before you begin

Reviewer: I feel like this section is kind-of a given (e.g. set up classroom) and unnecessary. @d12 what do you think?

Reviewer 2: I'm indifferent, I'll let y'all make a decision here.

Author: In most cases yes, but I added this in case someone is creating an assignment for the first time and visits this page for more information.

To use template repositories for assignments, you'll need the following:

* Set-up a GitHub classroom

> * An exercise (a repository that you have access to, which contains documentation, starter code, tests, or anything else your students need to begin work on an assignment).

Reviewer 2: If we keep this, this shouldn't be called an exercise since we use this word to refer to assignments. An exercise repo would be OK.

Author: I've taken this from here: https://classroom.github.com/help/create-group-assignments
To keep the language consistent, should we make this change in that doc as well?

Author: How about,

> An exercise repository that you have access to, which contains documentation, starter code, tests, or anything else your students need to begin work on an assignment.

#### Get started

> Once you log in to [GitHub Classroom](https://classroom.github.com/), select your classroom and then click on **New Assignment**. Template repositories work for both **Individual Assignments** and **Group Assignments**. So you can pick either of the two for your assignment. You can then provide details for the assignment. Once you've filled out the details, the **Optional** section has a search bar for finding your starter code repository. When you search for your starter code repository, you will now get two options:

Reviewer: `Template repositories work for both **Individual Assignments** and **Group Assignments**. So you can pick either of the two for your assignment.`

Second sentence seems repetitive/redundant

Author: Can you add a commit suggestion for this?  it's easier to update the doc from here 😄

> Once you log in to [GitHub Classroom](https://classroom.github.com/), select your classroom and then click on **New Assignment**. Template repositories work for both **Individual Assignments** and **Group Assignments**. So you can pick either of the two for your assignment. You can then provide details for the assignment. Once you've filled out the details, the **Optional** section has a search bar for finding your starter code repository. When you search for your starter code repository, you will now get two options:

Reviewer: ```suggestion
Once you log in to [GitHub Classroom](https://classroom.github.com/), select your classroom and then click on **New Assignment**. Template repositories work for both **Individual Assignments** and **Group Assignments**. You can then provide details for the assignment. Once you've filled out the details, the **Optional** section has a search bar for finding your starter code repository. When you search for your starter code repository, you will now get two options:
```

  * Import starter code using a template repository.
  * Import starter code using source importer.

If you choose the first option, we will use your starter code repository as a template for generating the assignment for the students.

![Searches for repository, then selects the option of Importing starter code with template repositories](https://user-images.githubusercontent.com/3170078/62091261-ee69f900-b224-11e9-96ab-b4b34cce0001.gif)

Note: For import via template repositories to work, your starter code repository _must_ be a template repository. If you own the repository or have permissions, you can do it manually from the settings page of the repository on GitHub. If you do not own the repository, you can make a fork of it and then make it a template. More details [here](https://help.github.com/en/articles/creating-a-template-repository). We are in the process of automating this for you, and will post an update on this page soon!

### What are the advantages of using template repositories?

* Importing starter code is faster, as it uses the new [template repositories API](https://developer.github.com/v3/repos/#create-repository-using-a-repository-template).

> * Template repositories are more reliable as we have noticed that 99% of our imports with template repositories succeeded, whereas only 95% of imports with source importer succeeded.

Reviewer 2: I know I just suggested we say template repos succeed 99% of the time, but reading now, I feel like a teacher of a class of 100 is going to read this as "1 student is going to get stuck every single assignment". Why don't we remove the percentages and go with this instead:

```suggestion
* Template repositories are far more reliable than source imports.
```

* A list of students, or unique identifiers like their email addresses.

#### Get started

Once you log in to [GitHub Classroom](https://classroom.github.com/), select your classroom and then click on **New Assignment**. Template repositories work for both **Individual Assignments** and **Group Assignments**. So you can pick either of the two for your assignment. You can then provide details for the assignment. Once you've filled out the details, the **Optional** section has a search bar for finding your starter code repository. When you search for your starter code repository, you will now get two options:

  * Import starter code using a template repository.
  * Import starter code using source importer.

If you choose the first option, we will use your starter code repository as a template for generating the assignment for the students.

![Searches for repository, then selects the option of Importing starter code with template repositories](https://user-images.githubusercontent.com/3170078/62091261-ee69f900-b224-11e9-96ab-b4b34cce0001.gif)

> Note: For import via template repositories to work, your starter code repository _must_ be a template repository.If you own the repository or have permissions, you can do it manually from the settings page of the repository on GitHub. More details [here](https://help.github.com/en/articles/creating-a-template-repository). We are in the process of automating this for you, and will post an update on this page soon!

Reviewer 2: Do we want to add something mentioning if they're using a repo they don't own, they can ask the owner to make it a template, or fork the repo and make that a template?

### What are the advantages of using template repositories?

> * Importing starter code is faster, as it uses the new [template repositories API](https://developer.github.com/v3/repos/#create-repository-using-a-repository-template).

Reviewer 2: This is great, but the main draw is reliability and error rate. You can quote some of the numbers from our datadog dashboard directly, 95% of importer imports succeed, 99.9% of template repo imports succeed

* Template repositories are built with the aim of distribution and management of boilerplate code and hence releasing assignments is almost instantaneous.

### What do you miss out on when you use template repositories?

Template repositories do not copy the complete commit history, but instead start with a single commit. Because of that, you miss out on the following features:

* Reference to all the original tags in the repositories.

> * Ability to push updates to the generated repositories.

Reviewer 2: Can we say "student repositories" here? I want to be _extra_ sure we don't imply they can't push updates to a template after making it a template.

### Recommended option for your assignment

Template repositories are the best option in _most_ cases. You may not want to use it if:

* You plan on pushing updates to the starter code repository _after_ releasing the assignment.
* You care about the full commit history of the starter code repository.

> Note: The template repositories API is still under a preview and might change.  We will keep this page updated if any of those changes affect your experience on classroom.

Reviewer 2: We should remove this note, the API being under preview means we won't have much time to adapt to upcoming changes, but that's our responsibility, not the end users. They can't do anything about this warning :)

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
```

### Helpful links:

> * [https://help.github.com/en/articles/creating-a-template-repository](https://help.github.com/en/articles/creating-a-template-repository)

Reviewer 3: ```suggestion
* [Creating a template repository](https://help.github.com/en/articles/creating-a-template-repository)
```

> * [https://help.github.com/en/articles/creating-a-repository-from-a-template](https://help.github.com/en/articles/creating-a-repository-from-a-template)

Reviewer 3: ```suggestion
* [Creating a repository from a template](https://help.github.com/en/articles/creating-a-repository-from-a-template)
```

> * [https://github.blog/2019-06-06-generate-new-repositories-with-repository-templates](https://github.blog/2019-06-06-generate-new-repositories-with-repository-templates)

Reviewer 3: ```suggestion
* [Generate new repositories with repository templates](https://github.blog/2019-06-06-generate-new-repositories-with-repository-templates)
```
